.github/workflows/mypy.yaml

```diff
@@ -0,0 +1,21 @@
+name: mypy
+​
+on: [push, pull_request]
+​
+jobs:
+  lint:
+    runs-on: ubuntu-latest
+​
+    steps:
+      - uses: actions/checkout@v2
+​
+      - name: Set up Python
+        uses: actions/setup-python@v1
+        with:
+          python-version: 3.7
+​
+      - name: Install
+        run: pip install mypy
+​
+      - name: Run linter
+        run: cd python && mypy -p lsst.daf.butler
```

Comment:
Is there some way to parameterize this so that it can be copied to other repos?

Reply:
Would `mypy python/lsst` work?

Reply:
People are reviewing this before I've even got it working 😄

Reply:
No, `cd python && mypy lsst` does not work at all. Maybe that's a misconfigured mypy.ini.

Reply:
I played around with this for quite a while before I got it working on the original travis version.  The problem is that mypy's support for "namespace packages" (directories that lack an `__init__.py`, as we do in our outermost `lsst` package) isn't great, and it doesn't provide any way to tell it what the root of the package tree is when there's no `__init__.py` to indicate it.  It defaults to assuming it's the current working directory, so we have to run it in such a way that `python` is the working directory.

Reply:
Oh, but I did not try `cd python && mypy lsst`; I think that stands a good chance of working.

Reply:
No:
``` $ mypy lsst
lsst/daf/__init__.py:2: error: Cannot determine type of '__path__'
lsst/daf/butler/registry/queries/exprParser/ply/lex.py:47: error: Module has no attribute "StringType"
lsst/daf/butler/registry/queries/exprParser/ply/lex.py:47: error: Module has no attribute "UnicodeType"
lsst/daf/butler/registry/queries/exprParser/ply/yacc.py:2002: error: Cannot find implementation or library stub for module named 'cPickle'
lsst/daf/butler/core/formatter.py:371: error: Too many values to unpack (2 expected, 3 provided)
lsst/daf/butler/core/formatter.py:372: error: Cannot determine type of 'formatter'
lsst/daf/butler/core/formatter.py:373: error: Cannot determine type of 'matchKey'
lsst/daf/butler/core/formatter.py:375: error: Cannot determine type of 'matchKey'
lsst/daf/butler/core/formatter.py:375: error: Cannot determine type of 'formatter'
lsst/daf/butler/core/formatter.py:371: error: Too many values to unpack (2 expected, 3 provided)
lsst/daf/butler/core/formatter.py:372: error: Cannot determine type of 'formatter'
lsst/daf/butler/core/formatter.py:373: error: Cannot determine type of 'matchKey'
lsst/daf/butler/core/formatter.py:375: error: Cannot determine type of 'matchKey'
lsst/daf/butler/core/formatter.py:375: error: Cannot determine type of 'formatter'
lsst/daf/butler/registry/_registry.py:81: error: Cannot find implementation or library stub for module named 'daf.butler.butlerConfig'
lsst/daf/butler/_deferredDatasetHandle.py:34: error: Cannot find implementation or library stub for module named 'daf.butler.butler'
lsst/daf/butler/_deferredDatasetHandle.py:88: error: Incompatible return value type (got "DataCoordinate", expected "ExpandedDataCoordinate")
lsst/daf/butler/_butler.py:209: error: Cannot determine type of '_config'
lsst/daf/butler/_butler.py:217: error: Incompatible types in assignment (expression has type "Union[bool, Tuple[str, ...]]", variable has type "Optional[bool]")
lsst/daf/butler/_butler.py:218: error: Argument "writeable" to "fromConfig" of "Registry" has incompatible type "Optional[bool]"; expected "bool"
lsst/daf/butler/_butler.py:380: error: Incompatible types in assignment (expression has type "ButlerURI", variable has type "str")
lsst/daf/butler/_butler.py:503: error: Incompatible return value type (got "Tuple[Optional[DatasetType], Union[DataCoordinate, Mapping[str, Any], None]]", expected "Tuple[DatasetType, Union[DataCoordinate, Mapping[str, Any]]]")
lsst/daf/butler/_butler.py:1087: error: Item "None" of "Optional[Iterable[str]]" has no attribute "__iter__" (not iterable)
lsst/daf/butler/_butler.py:1191: error: Need type annotation for 'resolvedRefs' (hint: "resolvedRefs: List[<type>] = ...")
lsst/daf/butler/_butler.py:1196: error: Need type annotation for 'allResolvedRefs' (hint: "allResolvedRefs: List[<type>] = ...")
lsst/daf/butler/_butler.py:1221: error: The return type of a generator function should be "Generator" or one of its supertypes
lsst/daf/butler/_butler.py:1221: error: Argument 1 to "contextmanager" has incompatible type "Callable[[Butler, DefaultNamedArg(Optional[str], 'directory'), DefaultNamedArg(Optional[str], 'filename'), DefaultNamedArg(Optional[str], 'format'), DefaultNamedArg(Optional[str], 'transfer')], ContextManager[RepoExport]]"; expected "Callable[..., Iterator[<nothing>]]"
lsst/daf/butler/_butler.py:1329: error: Value of type variable "AnyStr" of "splitext" cannot be "Union[str, TextIO]"
lsst/daf/butler/_butler.py:1329: error: Incompatible types in assignment (expression has type "Union[str, TextIO]", variable has type "Optional[str]")
lsst/daf/butler/_butler.py:1399: error: Argument 2 to "DatasetRef" has incompatible type "Dict[str, Any]"; expected "DataCoordinate"
lsst/daf/butler/_butler.py:1402: error: Argument 1 to "extend" of "list" has incompatible type "List[DatasetRef]"; expected "Iterable[DatasetType]"
lsst/daf/butler/_butler.py:1417: error: Incompatible types in assignment (expression has type "None", variable has type "DatasetType")
lsst/daf/butler/registry/tests/_registry.py:112: error: "RegistryTests" has no attribute "assertEqual"
lsst/daf/butler/registry/tests/_database.py:113: error: "DatabaseTests" has no attribute "assertCountEqual"
lsst/daf/butler/tests/_testRepo.py:26: error: Skipping analyzing 'numpy': found module but no type hints or library stubs
lsst/daf/butler/tests/_dummyRegistry.py:44: error: Need type annotation for '_rows' (hint: "_rows: List[<type>] = ...")
lsst/daf/butler/tests/_dummyRegistry.py:87: error: "DummyOpaqueTableStorageManager" has no attribute "_storage"; maybe "_storages"?
lsst/daf/butler/tests/_dummyRegistry.py:98: error: Need type annotation for '_bridges' (hint: "_bridges: Dict[<type>, <type>] = ...")
lsst/daf/butler/formatters/parquetFormatter.py:41: error: Skipping analyzing 'pyarrow.parquet': found module but no type hints or library stubs
lsst/daf/butler/formatters/parquetFormatter.py:41: note: See https://mypy.readthedocs.io/en/latest/running_mypy.html#missing-imports
lsst/daf/butler/formatters/parquetFormatter.py:41: error: Skipping analyzing 'pyarrow': found module but no type hints or library stubs
lsst/daf/butler/formatters/parquetFormatter.py:42: error: Skipping analyzing 'pandas': found module but no type hints or library stubs
lsst/daf/butler/formatters/parquetFormatter.py:96: error: Incompatible types in "yield" (actual type "Sequence[str]", expected type "Tuple[str, ...]")
lsst/daf/butler/formatters/parquetFormatter.py:146: error: Argument 1 to "_standardizeColumnParameter" of "_ParquetLoader" has incompatible type "Union[str, List[str], Dict[str, Union[str, List[str]]]]"; expected "Dict[str, Union[str, List[str]]]"
lsst/daf/butler/registry/datasets/byDimensions/_storage.py:54: error: Argument 1 of "insert" is incompatible with supertype "DatasetRecordStorage"; supertype defines the argument type as "RunRecord"
lsst/daf/butler/registry/datasets/byDimensions/_storage.py:90: error: Argument 1 of "find" is incompatible with supertype "DatasetRecordStorage"; supertype defines the argument type as "CollectionRecord"
lsst/daf/butler/registry/datasets/byDimensions/_storage.py:114: error: Argument 1 of "associate" is incompatible with supertype "DatasetRecordStorage"; supertype defines the argument type as "CollectionRecord"
lsst/daf/butler/registry/datasets/byDimensions/_storage.py:131: error: Argument 1 of "disassociate" is incompatible with supertype "DatasetRecordStorage"; supertype defines the argument type as "CollectionRecord"
lsst/daf/butler/registry/datasets/byDimensions/_storage.py:146: error: Argument 1 of "select" is incompatible with supertype "DatasetRecordStorage"; supertype defines the argument type as "CollectionRecord"
lsst/daf/butler/registry/datasets/byDimensions/_manager.py:100: error: Argument "db" to "ByDimensionsDatasetRecordStorage" has incompatible type "lsst.daf.butler.registry.interfaces._database.Database"; expected "daf.butler.registry.interfaces._database.Database"
lsst/daf/butler/registry/datasets/byDimensions/_manager.py:103: error: Argument "collections" to "ByDimensionsDatasetRecordStorage" has incompatible type "lsst.daf.butler.registry.interfaces._collections.CollectionManager"; expected "daf.butler.registry.interfaces._collections.CollectionManager"
lsst/daf/butler/registry/datasets/byDimensions/_manager.py:131: error: Argument "db" to "ByDimensionsDatasetRecordStorage" has incompatible type "lsst.daf.butler.registry.interfaces._database.Database"; expected "daf.butler.registry.interfaces._database.Database"
lsst/daf/butler/registry/datasets/byDimensions/_manager.py:134: error: Argument "collections" to "ByDimensionsDatasetRecordStorage" has incompatible type "lsst.daf.butler.registry.interfaces._collections.CollectionManager"; expected "daf.butler.registry.interfaces._collections.CollectionManager"
Found 52 errors in 14 files (checked 140 source files)
```

Reply:
It works fine if you put an empty `__init__.py` back at the top level.

Reply:
`cd python && touch lsst/__init__.py && mypy -p lsst`? :-)

(I'm not sure if that's a serious suggestion; it is hacky, but would mean we could use the same workflow ~everywhere which would be nice).

Reply:
Ok. Let me try that...

Reply:
That works fine.